File: solutions/scala/034-trapping-rain-water.scala
/**
 * Problem 34: Trapping Rain Water (LeetCode 42)
 * Difficulty: Hard
 * Language: Scala
 */
def trap(height: Array[Int]): Int = {
  val maxLeft = height.scanLeft(0)(_ max _).tail
  val maxRight = height.scanRight(0)(_ max _).init
  height.indices.map(i =>
    (maxLeft(i) min maxRight(i)) - height(i) max 0
  ).sum
}

File: solutions/haskell/034-trapping-rain-water.hs
{-
  Problem 34: Trapping Rain Water (LeetCode 42)
  Difficulty: Hard
  Language: Haskell
-}
trap :: [Int] -> Int
trap heights =
  let maxLeft  = scanl1 max heights
      maxRight = scanr1 max heights
      water = zipWith3 (\h l r -> max 0 (min l r - h)) heights maxLeft maxRight
  in sum water
